R/ibss.R

```diff
@@ -0,0 +1,182 @@
+# This file contains the three core functions for SuSiE (and its modifications):
+# 1) ibss_initialize(), 2) ibss_fit(), and 3) ibss_finalize(). All of these
+# functions make up the main workhorse "susie_engine()" function.
+
+
+### Initialization ###
+
+# This function initializes the SuSiE model object
+
+ibss_initialize <- function(data,
+                            L                     = min(10, ncol(data$X)),
+                            scaled_prior_variance = 0.2,
+                            residual_variance     = NULL,
+                            prior_weights         = NULL,
+                            null_weight           = NULL,
+                            model_init            = NULL){
+  # Define p
+  p <- data$p
+
+  # Check prior variance
+  if (!is.numeric(scaled_prior_variance) || scaled_prior_variance < 0)
+    stop("Scaled prior variance should be positive number")
+
+  # Check prior weights
+  if (is.null(prior_weights))
+    prior_weights <- rep(1 / p, p)
+
+  # Check number of single effects
+  if (p < L)
+    L = p
+
+  # Set variance of y
+  var_y <- if (!is.null(data$y)) var(drop(data$y)) else data$yty / (data$n - 1)
+
+  # Check residual variance
+  if (is.null(residual_variance)) {
+    residual_variance <- var_y
+  }
+
+  # Set alpha, mu, mu^2, V
+  alpha  <- matrix(1 / p, L, p)
+  mu     <- matrix(0,     L, p)
+  mu2    <- matrix(0,     L, p)
+  V      <- rep(scaled_prior_variance * var_y, L)
+
+  # Check for pre-initialized model
+  if (!missing(model_init) && !is.null(model_init)) {
+    if (!inherits(model_init, "susie"))
+      stop("model_init must be a 'susie' object")
+
+    if (nrow(model_init$alpha) > L)
+      stop("model_init has more effects than requested L")
+
+    if (max(model_init$alpha) > 1 || min(model_init$alpha) < 0)
+      stop("model_init$alpha has invalid values outside range [0,1]; please ",
+           "check your input")
+
+    # TODO: add more checks
+
+    model_init <- prune_zero_variance_effects(model_init)  # TODO: add generic here to prune effects with zero prior var
+
+    idx <- seq_len(nrow(model_init$alpha))
+    alpha[idx, ] <- model_init$alpha
+    mu[idx, ]    <- model_init$mu
+    mu2[idx, ]   <- model_init$mu2
+    V[idx]       <- model_init$V[idx]
+
+    if (!is.numeric(V))
+      stop("Input prior variance must be numeric")
+    if (!all(V >= 0))
+      stop("prior variance must be non-negative")
+    if (!all(dim(mu) == dim(mu2)))
+      stop("dimension of mu and mu2 in input object do not match")
+    if (!all(dim(mu) == dim(alpha)))
+      stop("dimension of mu and alpha in input object do not match")
+    if (nrow(alpha) != length(V))
+      stop("Input prior variance V must have length of nrow of alpha in ",
+           "input object")
+  }
+
+  Xr   <- if (inherits(data, "individual"))
+    susieR:::compute_Xb(data$X, colSums(alpha * mu)) else NULL
```

Comment:
we should not need `susieR:::` when we call internal functions. When you test the code you can always do this:

```
devtools::install()
```